### PR TITLE
Fix -Wformat issues

### PR DESCRIPTION
**Pull request by MichaelChirico**
Part of #5768

Still missing are any changes for `-Wformat-security`.

### CODE REVIEW

**Comment by MichaelChirico** [comment]
(PS it would be great if we could add these compiler warnings to CI to prevent regressions on this in the future)

**Comment by MichaelChirico** [comment]
I poked around in r-devel a bit to see how they handle `R_xlen_t` in `printf()` formatted messages, but didn't see anything.

Therefore the approach in 0d57fd8f9f793755a93d4e268cda7dec88f506b3 seems like the best expedient solution for now -- on most installations it will be a no-op (`long` -> `long` cast), and wherever `R_xlen_t` is just `int`, it'll cast `int` -> `long`.

**Comment by jangorecki** [comment]
> (PS it would be great if we could add these compiler warnings to CI to prevent regressions on this in the future)

If you would happen to that docker that reproduces that please share. It's probably just compiler version and flags but will make it easier to add.

**Comment by jangorecki** [comment]
I added `-Wformat -Wformat-security` to `~/.R/Makevars` and used gcc 13.2.0 (same as [CRAN's gcc](https://www.r-project.org/nosvn/R.check/r-devel-linux-x86_64-debian-gcc/data.table-00check.html)) but I am unable to reproduce those warnings from CRAN. Must be some extra flag then.

This is what we just talked about with @eddelbuettel about lack of CRAN reproducibility...

**Comment by MichaelChirico** [comment]
h/t @HughParsonage for this r-package-devel thread getting directly to the issue:

https://stat.ethz.ch/pipermail/r-package-devel/2023q4/010123.html

That points out R itself has a new macro `R_PRIdXLEN_T`:

https://github.com/r-devel/r-svn/blob/a9ded8dd775dc0aa75b26a5c88c11983d6d884f9/src/include/Rinternals.h#L75-L79

But we're miles away from being able to depend on such a recent addition to R. We could define the macro ourselves; it will be fragile but seems like no better choice at the moment if we want to avoid `long long`.

**Comment by jangorecki** [comment]
From the thread you linked it sounds `long long` is good enough, will work across different R versions.

**Comment by eddelbuettel** [comment]
I eyeballed the thread too but didn't like it much. I prefer to a) not depend on a particular version or b) make the code more complicated via an `#if ...` for a given architecture when c) a simple cast takes care of it.  So I am with @jangorecki here: a simple cast does it. I used `unsigned int` to convert a (C++_ `sizeof()` that made R on Windows bark. All quiet now.

**Comment by MichaelChirico** [comment]
thanks all. was only wary of `long long` given the comments in the CRAN release notes. Agree it's best otherwise.

**Comment by jangorecki** [comment]
Considering it was commented out in CRAN release we should be good

**Comment by jangorecki** [comment]
testing PR via
```
sudo docker run -it --rm rocker/r-devel /bin/bash
mkdir -p ~/.R
echo 'CFLAGS=-Wformat -Wformat-security' > ~/.R/Makevars
echo 'CXXFLAGS=-Wformat -Wformat-security' >> ~/.R/Makevars

## optional
wget https://cran.r-project.org/src/contrib/data.table_1.14.8.tar.gz
RD CMD INSTALL data.table_1.14.8.tar.gz
# warnings reproduced

## PR
wget https://github.com/Rdatatable/data.table/archive/refs/heads/format-warnings.zip
unzip format-warnings.zip
RD CMD build data.table-format-warnings --no-build-vignettes
RD CMD INSTALL data.table_1.14.9.tar.gz
```

File: src/assign.c [hunk 1]
```diff
@@ -468,7 +468,7 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values)
       // strong error message for now.
     else if (TRUELENGTH(names) != oldtncol)
       // Use (long long) to cast R_xlen_t to a fixed type to robustly avoid -Wformat compiler warnings, see #5768, PRId64 didnt work
-      error(_("Internal error: selfrefnames is ok but tl names [%ld] != tl [%ld]"), TRUELENGTH(names), oldtncol);  // # nocov
+      error(_("Internal error: selfrefnames is ok but tl names [%ld] != tl [%d]"), TRUELENGTH(names), oldtncol);  // # nocov
```
Comment (MichaelChirico) [inline src/assign.c:471]: oh, interesting, does that mean TRUELENGTH(names) and TRUELENGTH(dt) have different types??

I noticed the Wformat error only hit this line once & was wondering why.

Comment (jangorecki) [reply src/assign.c:471]: It appears to be like that. I didn't try to understand that.

**Pull request merged**